sound/soc/sof/ops.h

```diff
@@ -103,6 +103,22 @@ static inline int snd_sof_dsp_core_power_down(struct snd_sof_dev *sdev,
 	return ret;
 }
 
+static inline int snd_sof_dsp_core_get(struct snd_sof_dev *sdev, int core)
+{
+	if (sof_ops(sdev)->core_get)
+		return sof_ops(sdev)->core_get(sdev, core);
+
+	return 0;
+}
```

Comment:
Isn't this now overlapping with the power_up/down ops we already have?

Reply:
yeah it is a bit confusing but it is cleanest way to implement the notion of ref_counting for maintaining topology widgets.
The problem with the power_up/down ops mainly is the fact that they take a core_mask as input . So what's the problem with the core mask? Because it makes the function really ugly to write/maintain. I's have to go through each core in teh mask, power up/donw them and update the ref_count. And if one core fails, I'd to undo the cores that were already powered up/down. Trust me, I tried this and went nowhere.

For topology widgets, you'd never need to power up/down more than 1 core and it is so much cleaner to update the ref_count for just that one core in core_get/put. And it doesnt overlap with power up/down to be precise. It does more than power up/down byt powering up from the host and sending the IPC as done for CNL/ICL.

Reply:
@ranj063 But isn't that a property of Intel hw and we have no real need to have a core_mask in the ops-level interface? I mean the problem I have is that now we have two interfaces in ops to do the same thing, to power up a core. Whether the actual action is a MMIO write or IPC, that's an implementation detail, but for SOF core, there should be just one interface to power up/down. The whole host_managed_core is Intel hw detail that should be hidden in DSP implementation.

Reply:
> But isn't that a property of Intel hw and we have no real need to have a core_mask in the ops-level interface?

core_mask is usual in situatoins where we want to power off all cores like we do today in suspend/resume or even during FW loading for example. But I think what we should consider is having the power_up/down op for a single core and have a wrapper that can act of a core_mask for the suspend/remove/loading usecase. I will think about this.

Reply:
@kv2019i I had the same feedback and talked with @ranj063 on this, I also found that the concept of 'secondary core' was far from generic. The problem is that it's not "just" a "power-up a core" problem. We have also the CTX_SAVE IPC that needs to happen and different hardware-specific sequences, rather complicated I am afraid, to the point where I am wondering if we need to have a quick fix and completely redo the core power management/suspend/resume with a clean slate. That will be for after the holiday week-end :-)